File: app/src/main/java/com/mecong/tenderalarm/sleep_assistant/SleepAssistantFragment.java
package com.mecong.tenderalarm.sleep_assistant;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.graphics.PorterDuff;
import android.media.AudioManager;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.transition.ChangeBounds;
import android.transition.ChangeImageTransform;
import android.transition.Fade;
import android.transition.TransitionSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.viewpager.widget.ViewPager;

import com.google.android.material.tabs.TabLayout;
import com.hypertrack.hyperlog.HyperLog;
import com.mecong.tenderalarm.R;
import com.mecong.tenderalarm.model.PropertyName;
import com.mecong.tenderalarm.model.SQLiteDBHelper;
import com.mecong.tenderalarm.sleep_assistant.media_selection.SleepMediaType;
import com.mecong.tenderalarm.sleep_assistant.media_selection.SleepNoise;
import com.mecong.tenderalarm.sleep_assistant.media_selection.SoundListsPagerAdapter;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;

import java.util.concurrent.TimeUnit;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import lombok.AccessLevel;
import lombok.experimental.FieldDefaults;

import static com.mecong.tenderalarm.alarm.AlarmUtils.TAG;

@FieldDefaults(level = AccessLevel.PUBLIC)
public class SleepAssistantFragment extends Fragment {

    private static final long STEP_MILLIS = TimeUnit.SECONDS.toMillis(10);
    @BindView(R.id.pp_button)
    ImageButton ppButton;
    @BindView(R.id.textViewTime)
    TextView textViewTime;
    @BindView(R.id.viewPager)
    ViewPager viewPager;
    @BindView(R.id.tabs)
    TabLayout tabs;
    @BindView(R.id.textViewMinutes)
    TextView textViewMinutes;
    @BindView(R.id.nowPlayingText)
    TextView nowPlayingText;
    @BindView(R.id.textViewVolumePercent)
    TextView textViewVolumePercent;
    @BindView(R.id.slliderSleepTime)
    HourglassComponent sliderSleepTime;
    @BindView(R.id.sliderVolume)
    HourglassComponent sliderVolume;
    long timeMinutes;
    long timeMs;
    float volume;
    Handler handler;
    Runnable runnable;
    float volumeStep;
    boolean serviceBound;
    SleepAssistantPlayListModel playListModel;
    private RadioService service;
    ServiceConnection serviceConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName arg0, IBinder binder) {
            service = ((RadioService.LocalBinder) binder).getService();
            service.setAudioVolume(volume / 100f);
            serviceBound = true;
        }

        @Override
        public void onServiceDisconnected(ComponentName arg0) {
            serviceBound = false;
        }
    };


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        ViewGroup rootView = (ViewGroup) inflater.inflate(
                R.layout.content_sleep_assistant, container, false);

        ButterKnife.bind(this, rootView);

        HyperLog.i(TAG, "Sleep assistant fragment create view");

        if (savedInstanceState != null) {
            return rootView;
        }


        playListModel = (new ViewModelProvider(this)).get(SleepAssistantPlayListModel.class);

        if (playListModel.getPlaylist().getValue() == null) {
            SleepAssistantPlayListModel.PlayList defaultPlayList =
                    new SleepAssistantPlayListModel.PlayList(SleepNoise.Companion.retrieveNoises().get(0).getUrl(),
                            SleepNoise.Companion.retrieveNoises().get(0).getName(), SleepMediaType.NOISE);
            playListModel.setPlaylist(defaultPlayList);
        }

        playListModel.getPlaylist().observe(getViewLifecycleOwner(), new Observer<SleepAssistantPlayListModel.PlayList>() {
            @Override
            public void onChanged(SleepAssistantPlayListModel.PlayList playList) {
                if (service != null)
                    service.playMediaList(playList);
            }
        });


        final Context context = this.getContext();

        final SQLiteDBHelper instance = SQLiteDBHelper.getInstance(getContext());
        final Integer activeTab = instance.getPropertyInt(PropertyName.ACTIVE_TAB);
        initializeTabsAndMediaFragments(context, activeTab);

        timeMinutes = 39;

        AudioManager audioManager = (AudioManager) this.getActivity().getSystemService(Context.AUDIO_SERVICE);
        int streamMaxVolume = audioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC);
        int systemVolume = audioManager.getStreamVolume(AudioManager.STREAM_MUSIC);

        float volumeCoefficient = (float) systemVolume / streamMaxVolume;

        if (volumeCoefficient < 0.3f) {
            volumeCoefficient = 0.3f;
            audioManager.setStreamVolume(
                    AudioManager.STREAM_MUSIC, (int) (streamMaxVolume * volumeCoefficient), 0);
            Toast.makeText(context, "System volume set to 30%", Toast.LENGTH_SHORT).show();
        }
        volume = 105 - 100 * volumeCoefficient;


        sliderSleepTime.addListener(new HourglassComponent.SleepTimerViewValueListener() {
            @Override
            public void onValueChanged(long newValue) {
                textViewMinutes.setText(context.getString(R.string.sleep_minutes, newValue));
                timeMinutes = newValue;
                timeMs = TimeUnit.MINUTES.toMillis(timeMinutes);
                volumeStep = volume * STEP_MILLIS / timeMs;
            }
        });

        sliderVolume.addListener(new HourglassComponent.SleepTimerViewValueListener() {
            @Override
            public void onValueChanged(long newValue) {
                textViewVolumePercent.setText(context.getString(R.string.volume_percent, newValue));
                volume = newValue;
                service.setAudioVolume(volume / 100f);
                volumeStep = volume * STEP_MILLIS / timeMs;
            }
        });

        sliderSleepTime.setCurrentValue(timeMinutes);
        sliderVolume.setCurrentValue((long) volume);
        timeMs = TimeUnit.MINUTES.toMillis(timeMinutes);
        volumeStep = volume * STEP_MILLIS / timeMs;

        textViewMinutes.setText(context.getString(R.string.sleep_minutes, timeMinutes));
        textViewVolumePercent.setText(context.getString(R.string.volume_percent, (int) volume));


        handler = new Handler();

        runnable = new Runnable() {
            @Override
            public void run() {
                timeMs -= STEP_MILLIS;
                if (timeMs <= 0) {
                    handler.removeCallbacks(runnable);

                    if (service.isPlaying()) {
                        service.stop();
                        timeMinutes = 30;
                        sliderSleepTime.setCurrentValue(timeMinutes);
                        textViewMinutes.setText(context.getString(R.string.sleep_minutes, timeMinutes));

                        volume = 30;
                        service.setAudioVolume(volume / 100f);
                        sliderVolume.setCurrentValue((long) volume);
                        textViewVolumePercent.setText(context.getString(R.string.volume_percent, (int) volume));
                    }
                } else {
                    timeMinutes = TimeUnit.MILLISECONDS.toMinutes(timeMs);
                    sliderSleepTime.setCurrentValue(timeMinutes);
                    textViewMinutes.setText(context.getString(R.string.sleep_minutes, timeMinutes));

                    volume -= volumeStep;
                    service.setAudioVolume(volume / 100f);
                    sliderVolume.setCurrentValue((long) volume);
                    textViewVolumePercent.setText(context.getString(R.string.volume_percent, (int) volume));

                    handler.postDelayed(this, STEP_MILLIS);
                }
            }
        };

        return rootView;
    }

    private void initializeTabsAndMediaFragments(Context context, int activeTab) {
        SoundListsPagerAdapter soundListsPagerAdapter =
                new SoundListsPagerAdapter(this.getActivity().getSupportFragmentManager(), context, playListModel);
        viewPager.setAdapter(soundListsPagerAdapter);

        tabs.setupWithViewPager(viewPager);

        tabs.setSelectedTabIndicator(null);
        tabs.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {
            final SQLiteDBHelper instance = SQLiteDBHelper.getInstance(getContext());

            @Override
            public void onTabSelected(TabLayout.Tab tab) {
                tab.getCustomView().findViewById(R.id.view).setVisibility(View.INVISIBLE);
                tab.getCustomView().setBackground(getResources().getDrawable(R.drawable.tr2_background));

                instance.setPropertyString(PropertyName.ACTIVE_TAB, String.valueOf(tab.getPosition()));
            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {
                tab.getCustomView().findViewById(R.id.view).setVisibility(View.VISIBLE);
                tab.getCustomView().setBackground(getResources().getDrawable(R.drawable.tr1_background));
            }

            @Override
            public void onTabReselected(TabLayout.Tab tab) {
                onTabSelected(tab);
            }
        });

        for (int i = 0; i < tabs.getTabCount(); i++) {
            tabs.getTabAt(i).setCustomView(R.layout.media_tab);
        }

        ((ImageView) tabs.getTabAt(0).getCustomView().findViewById(R.id.imageView)).setImageResource(R.drawable.local_media);
        ((ImageView) tabs.getTabAt(1).getCustomView().findViewById(R.id.imageView)).setImageResource(R.drawable.online_media);
        ((ImageView) tabs.getTabAt(2).getCustomView().findViewById(R.id.imageView)).setImageResource(R.drawable.noises);

        tabs.getTabAt(activeTab % tabs.getTabCount()).select();
    }


    @OnClick(R.id.pp_button)
    void onPlayPauseButtonsClicked() {
        if (service.isPlaying()) {
            service.pause();
        } else {
            if (service.hasPlayList()) {
                service.resume();
            } else {
                service.playMediaList(playListModel.getPlaylist().getValue());
            }
        }
    }
//
//    private void acquireWakeLock() {
//        if (wakeLock == null) {
//            PowerManager powerManager = (PowerManager) this.getActivity().getSystemService(POWER_SERVICE);
//            wakeLock = powerManager.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK,
//                    "TenderAlarm::SleepAssistantWakelockTag");
//        }
//
//        if (wakeLock != null && !wakeLock.isHeld()) {
//            wakeLock.acquire(timeMs);
//        }
//    }
//
//    private void releaseWakeLock() {
//        if (wakeLock != null && wakeLock.isHeld()) {
//            wakeLock.release();
//        }
//    }

    @Subscribe(sticky = true)
    public void onPlayFileChanged(SleepAssistantPlayListModel.Media media) {
        nowPlayingText.setText(media.getTitle());
    }


    @Subscribe(sticky = true)
    public void onEvent(String status) {

        ppButton.setEnabled(true);
        switch (status) {
            case RadioService.LOADING:
                ppButton.setImageResource(R.drawable.pause_btn);
                ppButton.setColorFilter(0);

                handler.removeCallbacks(runnable);
                break;
            case RadioService.ERROR:
                nowPlayingText.setText(getString(R.string.can_not_stream));
                Toast.makeText(this.getContext(), getString(R.string.can_not_stream), Toast.LENGTH_SHORT).show();
                ppButton.setImageResource(R.drawable.play_btn);

                playListModel.playing.setValue(false);
                break;
            case RadioService.PLAYING:


                TransitionSet transition = new TransitionSet()
                        .addTransition(new ChangeBounds().setDuration(500))
                        .addTransition(new ChangeImageTransform().setDuration(500))
                        .addTransition(new Fade(Fade.OUT).setDuration(500));
//                TransitionManager.beginDelayedTransition((ViewGroup) findViewById(R.id.now_playing), transition);
                ppButton.setImageResource(R.drawable.pause_btn);
//                ppButton.setPaddingRelative(10,10,10,10);
                ppButton.setImageTintMode(PorterDuff.Mode.ADD);


                playListModel.playing.setValue(true);

                handler.removeCallbacks(runnable);
                handler.postDelayed(runnable, STEP_MILLIS);
                break;
            default:
                ppButton.setImageResource(R.drawable.play_btn);

                playListModel.playing.setValue(false);
                handler.removeCallbacks(runnable);
                break;
        }
//
//        trigger.setImageResource(status.equals(PlaybackStatus.PLAYING)
//                ? R.drawable.ic_pause_black
//                : R.drawable.ic_play_arrow_black);

    }

    @Override
    public void onStart() {
        super.onStart();
        HyperLog.i(TAG, "SA OnStart");
        bindRadioService();

        EventBus.getDefault().register(this);
    }

    @Override
    public void onStop() {
        super.onStop();
        HyperLog.i(TAG, "SA onStop");

        EventBus.getDefault().unregister(this);
    }

    @Override
    public void onPause() {
        super.onPause();
        HyperLog.i(TAG, "SA onPause");

    }

    @Override
    public void onDestroy() {
        EventBus.getDefault().unregister(this);

        handler.removeCallbacks(runnable);

        if (service.isPlaying()) {
            service.stop();
        }

        if (serviceBound) {
            this.getActivity().getApplication().unbindService(serviceConnection);
        }

        super.onDestroy();
    }

    @Override
    public void onResume() {
        super.onResume();
        HyperLog.i(TAG, "SA onResume");
    }

    private void bindRadioService() {
        Intent intent = new Intent(this.getContext().getApplicationContext(), RadioService.class);
        this.getActivity().getApplication().bindService(intent, serviceConnection, Context.BIND_AUTO_CREATE);

        if (service != null)
            EventBus.getDefault().post(service.getStatus());
    }

}

File: app/src/main/java/com/mecong/tenderalarm/sleep_assistant/SleepAssistantFragment.kt
package com.mecong.tenderalarm.sleep_assistant

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.media.AudioManager
import android.os.Bundle
import android.os.Handler
import android.os.IBinder
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.Toast
import androidx.core.content.res.ResourcesCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayout.OnTabSelectedListener
import com.mecong.tenderalarm.R
import com.mecong.tenderalarm.alarm.MainActivityMessages
import com.mecong.tenderalarm.databinding.ContentSleepAssistantBinding
import com.mecong.tenderalarm.model.PropertyName
import com.mecong.tenderalarm.model.SQLiteDBHelper.Companion.sqLiteDBHelper
import com.mecong.tenderalarm.sleep_assistant.RadioService.LocalBinder
import com.mecong.tenderalarm.sleep_assistant.media_selection.SleepMediaType
import com.mecong.tenderalarm.sleep_assistant.media_selection.SoundListsPagerAdapter
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import timber.log.Timber
import java.util.concurrent.TimeUnit
import kotlin.math.min
import kotlin.math.roundToInt
import kotlin.math.roundToLong

class SleepAssistantFragment : Fragment() {
  var volume = 0f
  var volumeStep = 0f
  var timeMs: Long = 0
  var timeMinutes: Long = 0
  private val sleepTimeHandler: Handler = Handler()
  private lateinit var sleepTimeRunnable: Runnable
  private val progressHandler: Handler = Handler()
  private lateinit var progressRunnable: Runnable
  private var showRadioBuffer = false

  var serviceBound = false
  lateinit var radioService: RadioService

  private var _binding: ContentSleepAssistantBinding? = null

  // This property is only valid between onCreateView and onDestroyView.
  private val binding get() = _binding!!

  private var serviceConnection: ServiceConnection = object : ServiceConnection {
    override fun onServiceConnected(arg0: ComponentName, binder: IBinder) {
      radioService = (binder as LocalBinder).service
      serviceBound = true
      radioService.audioVolume = volume / 100

      val dbHelper = sqLiteDBHelper(this@SleepAssistantFragment.requireContext())!!
      val savedShuffle = dbHelper.getPropertyString(PropertyName.SHUFFLE)?.toBoolean() ?: false
      radioService.shuffleModeEnabled = savedShuffle

      if (!EventBus.getDefault().isRegistered(this@SleepAssistantFragment)) {
        EventBus.getDefault().register(this@SleepAssistantFragment)
      }
    }

    override fun onServiceDisconnected(arg0: ComponentName) {
      serviceBound = false
    }
  }

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)

    playListModel = ViewModelProvider(this).get(SleepAssistantPlayListModel::class.java)

    val context = this.requireContext()
    val dbHelper = sqLiteDBHelper(context)!!

    showRadioBuffer = dbHelper.getPropertyString(PropertyName.SHOW_RADIO_BUFFER) == "1"

    binding.textViewMinutes.setOnLongClickListener {
      showRadioBuffer = !showRadioBuffer
      dbHelper.setPropertyString(PropertyName.SHOW_RADIO_BUFFER, if (showRadioBuffer) "1" else "0")
      true
    }

    val activeTab = dbHelper.getPropertyInt(PropertyName.ACTIVE_TAB) ?: 2
    initializeTabsAndMediaFragments(context, activeTab)

    timeMinutes = dbHelper.getPropertyLong(PropertyName.SLEEP_TIME) ?: 39
    binding.sliderSleepTime.setCurrentValue(timeMinutes)
    binding.textViewMinutes.text =
      context.resources.getQuantityString(R.plurals.n_minutes_plural, timeMinutes.toInt(), timeMinutes.toInt())

    binding.sliderSleepTime.addListener(object : SleepTimerViewValueListener {
      override fun onValueChanged(newValue: Long) {
        dbHelper.setPropertyString(PropertyName.SLEEP_TIME, newValue.toString())
        binding.textViewMinutes.text =
          context.resources.getQuantityString(R.plurals.n_minutes_plural, newValue.toInt(), newValue.toInt())
        timeMinutes = newValue
        timeMs = TimeUnit.MINUTES.toMillis(timeMinutes)
        volumeStep = volume * STEP_MILLIS / timeMs
      }
    })

    binding.sliderVolume.addListener(object : SleepTimerViewValueListener {
      override fun onValueChanged(newValue: Long) {
        binding.textViewVolumePercent.text = context.getString(R.string.volume_percent, newValue)
        volume = newValue.toFloat()
        radioService.audioVolume = volume / 100f
        volumeStep = volume * STEP_MILLIS / timeMs
      }
    })

    sleepTimeRunnable = Runnable {
      timeMs -= STEP_MILLIS
      if (timeMs <= 0) {
        sleepTimeHandler.removeCallbacks(sleepTimeRunnable)

        if (radioService.isPlaying) {
          radioService.pause()

          timeMinutes = 30
          binding.sliderSleepTime.setCurrentValue(timeMinutes)
          timeMs = TimeUnit.MINUTES.toMillis(timeMinutes)

          binding.textViewMinutes.text = context.resources.getQuantityString(
            R.plurals.n_minutes_plural,
            timeMinutes.toInt(),
            timeMinutes.toInt()
          )

          volume = 30f
          binding.sliderVolume.setCurrentValue(volume.toLong())
          binding.textViewVolumePercent.text = context.getString(R.string.volume_percent, volume.roundToInt())
          radioService.audioVolume = volume / 100f

          volumeStep = volume * STEP_MILLIS / timeMs
        }
      } else {
        timeMinutes = TimeUnit.MILLISECONDS.toMinutes(timeMs)
        binding.sliderSleepTime.setCurrentValue(timeMinutes)
        binding.textViewMinutes.text = context.resources.getQuantityString(
          R.plurals.n_minutes_plural,
          timeMinutes.toInt(),
          timeMinutes.toInt()
        )

        volume -= volumeStep
        radioService.audioVolume = volume / 100f
        binding.sliderVolume.setCurrentValue(volume.toLong())

        binding.textViewVolumePercent.text = context.getString(R.string.volume_percent, volume.roundToInt())

        sleepTimeHandler.postDelayed(sleepTimeRunnable, STEP_MILLIS)
      }
    }


    binding.nowPlayingText.setOnClickListener {
      val currentTab = dbHelper.getPropertyInt(PropertyName.ACTIVE_TAB) ?: 2
      binding.tabs.getTabAt(currentTab % binding.tabs.tabCount)!!.select()
    }
  }

  fun flipShuffleMode(): Boolean {
    radioService.shuffleModeEnabled = !radioService.shuffleModeEnabled
    return radioService.shuffleModeEnabled
  }


  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    bindRadioService()
  }

  override fun onCreateView(
    inflater: LayoutInflater, container: ViewGroup?,
    savedInstanceState: Bundle?
  ): View {
    _binding = ContentSleepAssistantBinding.inflate(inflater, container, false)
    return binding.root
  }

  private fun initializeTabsAndMediaFragments(context: Context?, activeTab: Int) {
    if (context == null) return

    val soundListsPagerAdapter = SoundListsPagerAdapter(this.activity?.supportFragmentManager, context)
    binding.viewPager.adapter = soundListsPagerAdapter
    binding.tabs.setupWithViewPager(binding.viewPager)
    binding.tabs.setSelectedTabIndicator(null)

    binding.tabs.addOnTabSelectedListener(object : OnTabSelectedListener {
      override fun onTabSelected(tab: TabLayout.Tab) {
        tab.customView?.findViewById<View>(R.id.blackLine)?.visibility = View.INVISIBLE
        tab.customView?.background = ResourcesCompat.getDrawable(resources, R.drawable.tr2_background, null)
      }

      override fun onTabUnselected(tab: TabLayout.Tab) {
        tab.customView?.findViewById<View>(R.id.blackLine)?.visibility = View.VISIBLE
        tab.customView?.background = ResourcesCompat.getDrawable(resources, R.drawable.tr1_background, null)
      }

      override fun onTabReselected(tab: TabLayout.Tab) {
        onTabSelected(tab)
      }
    })

    for (i in 0 until binding.tabs.tabCount) {
      binding.tabs.getTabAt(i)?.setCustomView(R.layout.media_tab)
    }

    (binding.tabs.getTabAt(0)?.customView?.findViewById<View>(R.id.imageView) as ImageView)
      .setImageResource(R.drawable.local_media)
    (binding.tabs.getTabAt(1)?.customView?.findViewById<View>(R.id.imageView) as ImageView)
      .setImageResource(R.drawable.online_media)
    (binding.tabs.getTabAt(2)?.customView?.findViewById<View>(R.id.imageView) as ImageView)
      .setImageResource(R.drawable.noises)

    binding.tabs.getTabAt(activeTab % binding.tabs.tabCount)!!.select()
  }

  @Subscribe
  fun onPlayButtonClicked(message: MainActivityMessages) {
    if (message == MainActivityMessages.SWITCH_PLAYBACK) {
      if (radioService.isPlaying) {
        radioService.pause()
      } else {
        if (radioService.hasPlayList()) {
          radioService.resume()
        } else {
          radioService.setMediaList(playListModel.playlist.value!!)
        }
      }
    }
  }


  @Subscribe
  fun onEvent(status: RadioServiceStatus) {
    when (status) {
      RadioServiceStatus.LOADING -> {
        sleepTimeHandler.removeCallbacks(sleepTimeRunnable)
      }
      RadioServiceStatus.ERROR -> {
        binding.nowPlayingText.text = getString(R.string.can_not_stream)
        Toast.makeText(this.context, getString(R.string.can_not_stream), Toast.LENGTH_SHORT).show()
        playListModel.playing.setValue(false)
      }
      RadioServiceStatus.PLAYING -> {

        startTimelineWatcher()
        playListModel.playing.value = true
        sleepTimeHandler.removeCallbacks(sleepTimeRunnable)
        sleepTimeHandler.postDelayed(sleepTimeRunnable, STEP_MILLIS)
      }
      else -> {
        playListModel.playing.value = false
        sleepTimeHandler.removeCallbacks(sleepTimeRunnable)
      }
    }
  }

  override fun onDestroy() {
    EventBus.getDefault().unregister(this)
    sleepTimeHandler.removeCallbacksAndMessages(null)
    progressHandler.removeCallbacksAndMessages(null)
    if (::radioService.isInitialized) {
      if (radioService.isPlaying) {
        radioService.stop()
      }
      if (serviceBound) {
        this.activity?.application?.unbindService(serviceConnection)
      }
    }
    super.onDestroy()
  }

  override fun onResume() {
    super.onResume()

    if (!::radioService.isInitialized || !radioService.isPlaying) {
      val audioManager = this.activity?.getSystemService(Context.AUDIO_SERVICE) as AudioManager
      val streamMaxVolume = audioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC)
      val systemVolume = audioManager.getStreamVolume(AudioManager.STREAM_MUSIC)
      var volumeCoefficient = systemVolume.toFloat() / streamMaxVolume
      if (volumeCoefficient >= 0.3f) {
        volumeCoefficient = 0.05f
        audioManager.setStreamVolume(
          AudioManager.STREAM_MUSIC, (streamMaxVolume * volumeCoefficient).roundToInt(), 0
        )
        Toast.makeText(
          this.requireActivity(),
          requireContext().getString(R.string.system_volume_toast),
          Toast.LENGTH_SHORT
        ).show()
      }

      volume = 105 - 100 * volumeCoefficient
      volume = min(volume, 100f)

      binding.sliderVolume.setCurrentValue(volume.toLong())
      timeMs = TimeUnit.MINUTES.toMillis(timeMinutes)
      volumeStep = volume * STEP_MILLIS / timeMs
      binding.textViewVolumePercent.text = this.activity?.getString(R.string.volume_percent, volume.roundToInt())
    }
  }

  private fun startTimelineWatcher() {
    val localFileProgress = Runnable {
      Timber.v(
        "Local: Buffered position %d, content %d",
        radioService.getBufferedPos(),
        radioService.getContentPos()
      )
      binding.nowPlayingText.setPlayPosition(radioService.getContentPos() / radioService.getContentDuration().toFloat())

      binding.playerTime1.text = getElapsedTime()
      binding.playerTime2.text = getTimeLeft()

      if (radioService.isPlaying) {
        progressHandler.postDelayed(progressRunnable, 300)
      }
    }

    val onlineStreamProgress = Runnable {
      Timber.v(
        "Online: Buffered position %d, content %d",
        radioService.getBufferedPos(),
        radioService.getContentPos()
      )

      binding.playerTime1.text = getElapsedTime()
      binding.playerTime2.text = getBufferedSize()

      if (radioService.sleepAssistantPlayList?.mediaType == SleepMediaType.ONLINE) {
        progressHandler.postDelayed(progressRunnable, 800)
      }
    }

    progressHandler.removeCallbacksAndMessages(null)
    when (radioService.sleepAssistantPlayList?.mediaType) {
      SleepMediaType.LOCAL -> {
        progressRunnable = localFileProgress
        progressHandler.post(progressRunnable)
        binding.playerTime2.visibility = View.VISIBLE
        binding.playerTime1.visibility = View.VISIBLE
      }
      SleepMediaType.ONLINE -> {
        if (showRadioBuffer) {
          progressRunnable = onlineStreamProgress
          progressHandler.post(progressRunnable)
          binding.playerTime2.visibility = View.VISIBLE
          binding.playerTime1.visibility = View.VISIBLE
        } else {
          binding.playerTime2.visibility = View.GONE
          binding.playerTime1.visibility = View.GONE
        }
      }
      else -> {
        binding.playerTime2.visibility = View.GONE
        binding.playerTime1.visibility = View.GONE
      }
    }
  }

  private fun getStringForTime(timeMs: Long, negative: Boolean): String? {
    val totalSeconds = if (timeMs < 0) 0 else (timeMs + 500) / 1000

    val seconds = totalSeconds % 60
    val minutes = totalSeconds / 60 % 60

    val formattedTime = if (totalSeconds > 3600) {
      this.context?.getString(R.string.play_time_with_hour, totalSeconds / 3600, minutes, seconds)
    } else {
      this.context?.getString(R.string.play_time, minutes, seconds)
    }

    return if (negative) {
      "-$formattedTime"
    } else
      formattedTime
  }

  private fun getElapsedTime(percentPos: Float? = null): CharSequence? {
    return if (percentPos == null) {
      getStringForTime(radioService.getContentPos(), false)
    } else {
      getStringForTime((percentPos * radioService.getContentDuration()).roundToLong(), false)
    }
  }

  private fun getTimeLeft(percentPos: Float? = null): CharSequence? {
    return if (percentPos == null) {
      getStringForTime(radioService.getContentDuration() - radioService.getContentPos(), true)
    } else {
      getStringForTime(
        radioService.getContentDuration() - (percentPos * radioService.getContentDuration()).roundToLong(),
        true
      )
    }
  }

  private fun getBufferedSize(): CharSequence? {
    val buff = radioService.getBufferedPos() - radioService.getContentPos()
    return getStringForTime(buff, true)
  }

  @Subscribe(sticky = true)
  fun onPlayFileChanged(playList: SleepAssistantPlayListActive) {
    binding.nowPlayingText.interactiveMode = playList.mediaType == SleepMediaType.LOCAL
    radioService.setMediaList(playList)
    playListModel.playlist.value = playList

    radioService.play()
  }

  @Subscribe(sticky = true)
  fun onPlayFileChanged(playList: SleepAssistantPlayListIdle) {
    binding.nowPlayingText.interactiveMode = playList.mediaType == SleepMediaType.LOCAL
    radioService.setMediaList(playList)
    playListModel.playlist.value = playList
  }

  @Subscribe
  fun onPlayPosChanged(playPosition: PlayPosition) {
    if (playPosition.final) {
      radioService.seekToPercent(playPosition.playPositionPercent)
      progressHandler.post(progressRunnable)
    } else {
      progressHandler.removeCallbacks(progressRunnable)
    }

    binding.playerTime2.text = getTimeLeft(playPosition.playPositionPercent)
    binding.playerTime1.text = getElapsedTime(playPosition.playPositionPercent)
  }

  @Subscribe
  fun onPlayFileChanged(media: Media) {
    binding.nowPlayingText.text = media.title
  }

  @Subscribe
  fun persistMediaPosition(playList: SleepAssistantPlayList) {
    if (this.context == null) return

    val dbHelper = sqLiteDBHelper(this.requireContext())!!

    val activeTab = when (playList.mediaType) {
      SleepMediaType.LOCAL -> "0"
      SleepMediaType.ONLINE -> "1"
      SleepMediaType.NOISE -> "2"
    }

    dbHelper.setPropertyString(PropertyName.ACTIVE_TAB, activeTab)
    dbHelper.setPropertyString(PropertyName.TRACK_NUMBER, playList.index.toString())
    dbHelper.setPropertyString(PropertyName.PLAYLIST_ID, playList.playListId.toString())
    dbHelper.setPropertyString(PropertyName.POSITION_IN_TRACK, radioService.getContentPos().toString())
  }

  private fun bindRadioService() {
    val intent = Intent(this.context, RadioService::class.java)
    this.requireActivity().application.bindService(intent, serviceConnection, Context.BIND_AUTO_CREATE)
  }

  companion object {
    private val STEP_MILLIS = TimeUnit.SECONDS.toMillis(10)
    lateinit var playListModel: SleepAssistantPlayListModel
  }
}